Allow creating blank (empty) wallets (alternative)

achow101 opened the pull request:
Alternative (kind of) to #14938

This PR adds a `blank` parameter to the `createwallet` RPC to create a wallet that has no private keys initially. `sethdseed` can then be used to make a clean wallet with a custom seed. `encryptwallet` can also be used to make a wallet that is born encrypted.

Instead of changing the version number as done in #14938, a wallet flag is used to indicate that the wallet should be blank. This flag is set at creation, and then unset when the wallet is no longer blank. A wallet becomes non-blank when a HD seed is set or anything is imported. The main change to create a blank wallet is primarily taken from #14938.

Also with this, the term "blank wallet" is used instead of "empty wallet" to avoid confusion with wallets that have balance which would also be referred to as "empty".

This is built on top of #15225 in order to fix GUI issues.

jonasschnelli commented on src/wallet/wallet.h:
```diff
@@ -137,9 +137,12 @@ enum WalletFlags : uint64_t {
     WALLET_FLAG_DISABLE_PRIVATE_KEYS = (1ULL << 32),
+
+    // Will enforce that the wallet should be blank
+    WALLET_FLAG_BLANK_WALLET = (1ULL << 33),
```
Should we mention in the release notes that blank wallets can not be used in older bitcoin core versions?

jonasschnelli commented:
Concept ACK

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -700,8 +710,8 @@ bool CWallet::EncryptWallet(const SecureString& strWalletPassphrase)
-        // if we are using HD, replace the HD seed with a new one
-        if (IsHDEnabled()) {
+        // if we are using HD, create a fresh seed or replace the existing seed with a new one
+        if (IsHDEnabled() && !IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS)) {
```
No. Looking at what we currently do, this should actually be checking for `HasHDSeed()`. Will change.

Sjors replied:
So this would add a seed to an empty wallet? That should be made more clear in the documentation.

What if I want to import private keys and encrypt those? Should we add a `bool` to `encryptwallet` to determine if it should generate a key?

Sjors commented:
Concept ACK, thanks for saving me some work :-)

Are you sure older clients will understand this mandatory flag? Testing this is a good use of #12134, though such test should be in a separate PR to prevent a long wait.

Sjors commented on src/wallet/wallet.cpp:
```diff
@@ -1415,11 +1416,25 @@ void CWallet::SetHDChain(const CHDChain& chain, bool memonly)
+    LOCK(cs_wallet);
+    // -usehd was removed in 0.16, so all wallets with FEATURE_PRE_SPLIT_KEYPOOL are HD
+    // For older wallets the presence of seed_id indicates if HD is enabled.
+    // Support for empty HD capable wallets was added in 0.18
```
Nit: "blank capable".

Sjors commented on src/wallet/wallet.cpp:
```diff
@@ -1415,11 +1416,25 @@ void CWallet::SetHDChain(const CHDChain& chain, bool memonly)
+
+bool CWallet::IsBlank() const
+{
+    return IsHDEnabled() && !HasHDSeed();
```
Nit: document that for older wallets this relies on `IsHDEnabled()` checking `HasHDSeed()`, so there's no way this returns a false positive.

ryanofsky commented:
This should probably replace #14938 on the high priority list: https://github.com/bitcoin/bitcoin/projects/8

Sjors reviewed (commented):
Other than my questions and remarks c0bfc2d7160e7c861d9a3e678e1dd9b6213a8a86 looks good.

achow101 commented:
> Are you sure older clients will understand this mandatory flag? Testing this is a good use of #12134, though such test should be in a separate PR to prevent a long wait.

Just tested with 0.17 and 0.16. 0.17 knows what the wallet flags are so it does understand the mandatory flag and refuses to load wallets with the blank flag set. However 0.16 does not know about wallet flags and will actually load the wallet and generate keys for it. It does this for both wallets with private keys disabled and blank wallets, which might be an issue. However I noticed it is a bit harder to move back to 0.16 due to the wallet directory change where wallets are now directories instead of just files. But this is an issue in general for wallet flags.

jonasschnelli commented:
Would it had made (or still make) sense to raise the wallet version number to prevent ignoring wallet flags?

ryanofsky commented:
> Would it had made (or still make) sense to raise the wallet version number to prevent ignoring wallet flags?

It makes sense that if you are going to set a mandatory flag, you also need to increase the minimum version to whatever bitcoin version started checking mandatory flags.

Sjors commented:
Wallet feature flags were introduced in 0.17.0, and as @achow101 and the source comment points out, "wallet flags in the upper section (> 1 << 31) will lead to not opening the wallet if flag is unknown".

The following, added in a8da482a8bc87ff26194612727d4a7b86b2fb60d and released in v0.17.0, should already prevent 0.16 nodes from opening a >=0.17.0 wallet:

```cpp
FEATURE_PRE_SPLIT_KEYPOOL = 169900,
FEATURE_LATEST = FEATURE_PRE_SPLIT_KEYPOOL;
```

@achow101 I'm surprised you were able to open such a wallet in 0.16. Are you sure you compiled it right? I just tried using a binary and it refused to open a wallet with `WALLET_FLAG_DISABLE_PRIVATE_KEYS`, because the version was too new.

By the way, let's make sure we don't get in the way of #13756 which introduces `MUTABLE_WALLET_FLAGS`, flags that once set can't be changed.

achow101 commented:
Oh, I made a mistake. Just tested it again and yes, 0.16 does not open wallets that have flags set, so there is no need to do another version bump.

laanwj commented:
> Also with this, the term "blank wallet" is used instead of "empty wallet" to avoid confusion with wallets that have balance which would also be referred to as "empty".

Thank you, I really like this distinct terminology.

laanwj commented on test/functional/test_framework/test_node.py:
```diff
@@ -434,7 +434,7 @@ def batch(self, requests):
         """Run bitcoin-cli command. Deserializes returned string as python object."""
         pos_args = [str(arg).lower() if type(arg) is bool else str(arg) for arg in args]
-        named_args = [str(key) + "=" + str(value) for (key, value) in kwargs.items()]
+        named_args = [str(key) + "=" + (str(value).lower() if type(value) is bool else str(value)) for (key, value) in kwargs.items()]
```
Same comment here as https://github.com/bitcoin/bitcoin/pull/15235#discussion_r250547555

achow101 replied:
Since this issue affects multiple PRs, I've split it into a separate PR: #15301

achow101 commented:
Rebased

Sjors commented on src/interfaces/wallet.cpp:
```diff
@@ -457,7 +457,7 @@ class WalletImpl : public Wallet
     }
     unsigned int getConfirmTarget() override { return m_wallet.m_confirm_target; }
-    bool hdEnabled() override { return m_wallet.IsHDEnabled(); }
+    bool hdEnabled() override { return m_wallet.HasHDSeed(); }
```
Let's fix `hdEnabled` in `WalletModel` instead, otherwise things are going to get confusing fast.

achow101 replied:
I've changed it so that `hdEnabled()` does the same thing as before and instead added a `hasHDSeed()` which calls `HasHDSeed`. `hasHDSeed` is used instead of `hdEnabled` in the `walletmodel.cpp` changes.

Sjors reviewed (commented):
2021dfa is getting close. It just needs some changes to prevent even more confusion around how we're deciding if a wallet is HD.

Also regarding encrypting a blank wallet, whatever strategy you pick should be documented. See my earlier comment:

> So this would add a seed to an empty wallet? That should be made more clear in the documentation.

> What if I want to import private keys and encrypt those? Should we add a `bool` to `encryptwallet` to determine if it should generate a key?

achow101 commented on test/functional/wallet_createwallet.py:
```diff
@@ -59,5 +59,27 @@ def run_test(self):
         w3.getnewaddress()
         w3.getrawchangeaddress()
 
+        self.log.info("Test blank creation with privkeys enabled and then encryption")
```
No, the order is correct. Github is just showing it in the wrong order because of the dates. I'll fix the dates so it shows correctly.

achow101 commented:
I've updated the docs and added a few code comments as suggested by earlier comments (which I can't respond to for some reason).

Sjors commented on src/qt/walletmodel.cpp:
```diff
@@ -581,11 +581,12 @@ bool WalletModel::privateKeysDisabled() const
-    //     * !IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS): private keys have not been disabled (which results in hdEnabled() == true)
-    return m_wallet->hdEnabled() || (!m_wallet->hdEnabled() && !m_wallet->IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS));
+    //     * !hasHDSeed(): an HD seed is not present; and
+    //     * !IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS): private keys have not been disabled (which results in hasHDSeed() == true)
```
nit: missing `; and`

Sjors commented on src/wallet/wallet.h:
```diff
@@ -1129,9 +1129,15 @@ class CWallet final : public CCryptoKeyStore, public CValidationInterface
     /* Returns true if HD is enabled */
     bool IsHDEnabled() const;
 
+    /* Returns true if the wallet is blank */
```
nit: `, i.e. it has no private keys` (because it might have scripts, and the keypool isn't taken into account)

Sjors reviewed (approved):
utACK 24ff3c7

achow101 commented:
Rebased

Sjors commented:
tACK 4d2b8b4, since last check: rebase and moved some tests to another commit

I checked again that v0.17.1 refuses to open a blank wallet, and that after `sethdseed`  it does.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -3295,6 +3311,9 @@ bool CWallet::TopUpKeyPool(unsigned int kpSize)
             return false;
 
+        // If wallet supports HD but no seed is present, do not top up keypool
+        if (IsBlank()) return true;
```
In commit "[wallet] support creating a blank wallet" (96476657a06797913edffd42d3fdae320b86e9df):

This isn't checking the WALLET_FLAG_BLANK_WALLET flag at all. It seems weird that to allow topping up keypool or generating keys (in GetKeyFromPool) in a wallet that supports hd but intentionally has no seed. Am I understanding this correctly, or is there another check somewhere else?

jonasschnelli replied:
I can't follow your comment.
The `IsBlank()` check here stops the refill-process if the wallet is either blank or if no HD seed is available (while HD is enabled).
I also tested this by creating a blank wallet, import a watch-only and then calling `keypoolrefill`.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/15226#discussion_r254514195

> I can't follow your comment.

This comment was wrong about the keypool, but it was added reviewing a previous version of this PR (commit hash mentioned was 96476657a06797913edffd42d3fdae320b86e9df) before isBlank checked the blank flag, which is probably why you were confused.

I actually think now that this shouldn't directly be checking the blank flag, but I want to look more closely and follow up with an other review.

ryanofsky commented on src/interfaces/wallet.cpp:
```diff
@@ -463,6 +463,7 @@ class WalletImpl : public Wallet
     }
     unsigned int getConfirmTarget() override { return m_wallet.m_confirm_target; }
     bool hdEnabled() override { return m_wallet.IsHDEnabled(); }
```
In commit "Set a wallet flag for blank wallets" (4d2b8b4d93cb979fba81182258d93d562608a632)

I don't get understand the GUI part of this change at all. Does it actually make sense to show an HD enabled logo for a wallet that doesn't have an HD key? And will generate non-HD keys if an address is requested (IIUC)?

ryanofsky reviewed (commented):
I really like the feature, but would NACK the PR in it's current form just because it so pointlessly confusing to review! The worst part is the renaming of IsHDEnabled to HasHDSeed followed by an addition of a new IsHDEnabled with a different meaning. So all the calls where the previous behavior isn't changing show up in the diff, while all the calls where the behavior **is** changing don't show up at all! I'd really suggest dropping the `IsHDEnabled` rename or doing a scripted diff commit, and then adding a new function like `IsHDSupported` or `IsHDSafe` for the new version-based check.

achow101 committed:
[wallet] Support creating a blank wallet

A blank wallet is a wallet that has no keys, script or watch only things.
A new wallet flag indicating that it is blank will be set when the wallet
is blank. Once it is no longer blank (a seed has been generated, keys or
scripts imported, etc), the flag will be unset.

achow101 commented:
I've reworked `IsHDEnabled()`, `HasHDSeed()`, and `IsBlank()`. I have decided to lave `IsHDEnabled()` as it is and dropped `HasHDSeed()`. Instead `IsBlank()` will now check whether the blank wallet flag is set or if HD support is specified in the version but no HD seed is set. This change should clear up any confusion.

I've also squashed down the commits as I don't think the original commit structure would pass tests.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -1418,6 +1429,13 @@ bool CWallet::IsHDEnabled() const
+{
+    // A Blank wallet either has the flag set or it has a version number requiring HD but no HD seed is set
+    LOCK(cs_wallet);
+    return IsWalletFlagSet(WALLET_FLAG_BLANK_WALLET) || (CanSupportFeature(FEATURE_HD) && !IsHDEnabled());
```
In commit "[wallet] Support creating a blank wallet" (1cdb8d8346649434450e410971cc3d68c3504a25)

If there a wallet was created after the FEATURE_HD version but with `-usehd=0` set, would it cause this function to return true and no longer let GetKeyFromPool and TopUpKeyPool work?

achow101 replied:
I don't think so because those wallets would have a version less than FEATURE_HD.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -374,7 +380,11 @@ bool CWallet::AddWatchOnly(const CScript& dest)
     NotifyWatchonlyChanged(true);
-    return WalletBatch(*database).WriteWatchOnly(dest, meta);
+    if (WalletBatch(*database).WriteWatchOnly(dest, meta)) {
+        UnsetWalletFlag(WALLET_FLAG_BLANK_WALLET);
```
In commit "[wallet] Support creating a blank wallet" (1cdb8d8346649434450e410971cc3d68c3504a25)

It looks like if you import a watch-only key into a blank wallet WalletModel::canGetAddresses() will start returning true. I wonder if canGetAddresses should be using IsBlank.

achow101 replied:
I think it should. I've added `isBlank()` to the wallet interface and `canGetAddresses()` calls it instead of checking the wallet flag.

ryanofsky reviewed (commented):
Thanks for reworking and squashing. I think I understand the PR better now. I do have two questions below.

jonasschnelli commented:
Tested ACK 8061ea4611696b22d5d5dcdd73c1e053ffecc09d

gwillen commented:
Tested on OS X, reviewed and looks good but I don't feel qualified to ACK as I don't understand the intricacies of the wallet well enough.

Sjors commented:
I'll try to give this another review in the next few days.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -1418,6 +1429,13 @@ bool CWallet::IsHDEnabled() const
+{
+    // A Blank wallet either has the flag set or it has a version number requiring HD but no HD seed is set
+    LOCK(cs_wallet);
+    return IsWalletFlagSet(WALLET_FLAG_BLANK_WALLET) || (CanSupportFeature(FEATURE_HD) && !IsHDEnabled());
```
In commit "[wallet] Support creating a blank wallet" (8061ea4611696b22d5d5dcdd73c1e053ffecc09d)

I think there is no point to checking WALLET_FLAG_BLANK_WALLET here. It's redundant and just complicates the logic.

Also, I don't think this function should be called IsBlank, because this creates two slightly different definitions of blank, one where a wallet stops being blank when a watch-only address is added (WALLET_FLAG_BLANK_WALLET), and one which remains blank (IsBlank). I think a more accurate name for this function would be CannotGenerateKeys, but even better would be:

```c++
bool CWallet::CanGenerateAddresses()
{
    // Wallet can generate addresses if it has an HD seed, or is a pre-hd wallet.
    return IsHDEnabled() || !CanSupportFeature(FEATURE_HD);
}
```

which could simplify all the places where this is called, as well as the corresponding GUI function.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/15226#discussion_r255249959

After seeing #14075, I'd tweak my suggestion above and call this `CanGenerateKeys` or `CanTopUpKeyPool` instead of `CanGenerateAddresses`. After #14075 it looks like it will be possible to generate addresses with imported keys even if the wallet can't generate new keys.

ryanofsky commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3880,7 +3891,7 @@ UniValue sethdseed(const JSONRPCRequest& request)
 
     // Do not do anything to non-HD wallets
-    if (!pwallet->IsHDEnabled()) {
+    if (!pwallet->IsBlank() && !pwallet->IsHDEnabled()) {
```
In commit "[wallet] Support creating a blank wallet" (8061ea4611696b22d5d5dcdd73c1e053ffecc09d)

This is pretty convoluted. It would be more straightforward to avoid all the cancelling conditions and just write `if (!CanSupportFeature(FEATURE_HD))`

ryanofsky replied:
In commit "[wallet] Support creating a blank wallet" (8061ea4611696b22d5d5dcdd73c1e053ffecc09d)

This will now allow calling `sethdseed` on a wallet with `WALLET_FLAG_DISABLE_PRIVATE_KEYS` set. That seems like a mistake because I'm not sure the resulting state would make sense, and at the very least `GenerateNewSeed` would assert false in this case.

Would suggest adding a check for WALLET_FLAG_DISABLE_PRIVATE_KEYS here.

achow101 replied:
There's already a check for that flag above.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -3416,7 +3446,7 @@ void CWallet::ReturnKey(int64_t nIndex, bool fInternal, const CPubKey& pubkey)
 bool CWallet::GetKeyFromPool(CPubKey& result, bool internal)
 {
-    if (IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS)) {
+    if (IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS) || IsBlank()) {
```
In commit "[wallet] Support creating a blank wallet" (8061ea4611696b22d5d5dcdd73c1e053ffecc09d)

Would suggest replacing this with if `(!CanGenerateAddresses()) return false` if you use my other suggestion.

ryanofsky commented on src/wallet/wallet.h:
```diff
@@ -136,9 +136,12 @@ enum WalletFlags : uint64_t {
     // will enforce the rule that the wallet can't contain any private keys (only watch-only/pubkeys)
     WALLET_FLAG_DISABLE_PRIVATE_KEYS = (1ULL << 32),
+
+    // Will enforce that the wallet should be blank
```
In commit "[wallet] Support creating a blank wallet" (8061ea4611696b22d5d5dcdd73c1e053ffecc09d)

It would be nice to explain how the flag works. I'd suggest.

```c++
//! Flag set when a wallet contains no HD seed and no private keys, scripts,
//! addresses, and other watch only things, and is therefore "blank."
//!
//! The only function this flag has is to distinguish a blank wallet from
//! a newly created wallet when the wallet database is loaded, to avoid
//! initialization that should only happen on first run.
//!
//! This flag is also a mandatory flag to prevent previous versions of
//! bitcoin from opening the wallet, thinking it is newly created, and
//! improperly initializing it.
```

ryanofsky reviewed (commented):
This is pretty good and I think it will work but I think it can be simplified more. Left suggestions below.

ryanofsky commented on src/qt/walletmodel.cpp:
```diff
@@ -581,11 +581,12 @@ bool WalletModel::privateKeysDisabled() const
+    //     * !hdEnabled(): a HD seed is not present; and
+    //     * !IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS): private keys have not been disabled
+    //     * !isBlank(): the wallet is not blank
+    return m_wallet->hdEnabled() || (!m_wallet->hdEnabled() && !m_wallet->IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS) && !m_wallet->isBlank());
```
In commit "[wallet] Support creating a blank wallet" (8061ea4611696b22d5d5dcdd73c1e053ffecc09d)

I think this entire line is equivalent to `return !m_wallet->isBlank()`.

achow101 commented:
I have taken @ryanofsky's suggestions and reworked this a bit more. This should also help with #14075.

I got rid of `IsBlank` and replaced it with two functions: `CanGenerateKeys` and `CanGetAddresses`. `CanGenerateKeys` is as @ryanofsky commented; it checks for whether there is an HD seed or whether the wallet is non-HD. `CanGetAddresses` checks whether there are keys in the keypool (defaults to external keypool but can check the internal one), and if not, checks `CanGenerateKeys`. This is used as a guard wherever a key from the keypool is fetched for addresses. Additionally this is called by `WalletModel::CanGetAddresses()` and replaces the logic that was there.

ryanofsky reviewed (commented):
utACK 8b5533ee281ba114b8bb5c195f4a9b715b02a644. Thanks for being so responsive. I think this is in good shape, though I did leave some more suggestions.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -3416,7 +3460,7 @@ void CWallet::ReturnKey(int64_t nIndex, bool fInternal, const CPubKey& pubkey)
 bool CWallet::GetKeyFromPool(CPubKey& result, bool internal)
 {
-    if (IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS)) {
+    if (!CanGetAddresses(internal)) {
```
The reason I have the `CanGetAddresses` check here is for future work where you can get a key from the keypool without the ability to generate addresses.

Sjors replied:
Agree with the usefulness of `CanGetAddresses` for followup stuff. Longer term I also think it's a useful distinction in a descriptor based wallet, which may support a lot more (semi) watch-only (solvable, but not IsMine) things like being part of a multisig group.

 Regarding the "racy" stuff, maybe it helps to add `AssertLockHeld` to `CWallet CanGetAddresses` and `CanGenerateKeys`? That way places that (indirectly) call both can get a lock once.

achow101 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -176,6 +176,11 @@ static UniValue getnewaddress(const JSONRPCRequest& request)
 
     LOCK(pwallet->cs_wallet);
 
+    if (!pwallet->CanGetAddresses()) {
```
I've added a comment

MeshCollider commented:
utACK https://github.com/bitcoin/bitcoin/pull/15226/commits/707eabab827250cc1b726179cabe97d3e0c100de

I'll give @Sjors a chance to finish his last review of this, then merge tomorrow

Sjors commented on src/wallet/wallet.h:
```diff
@@ -1132,6 +1144,12 @@ class CWallet final : public CCryptoKeyStore, public CValidationInterface
     /* Returns true if HD is enabled */
```
`IsHDEnabled()` should be renamed to `HasHDSeed()` because that's actually what it checks (as opposed to `CanSupportFeature(FEATURE_HD)`). Can perhaps wait until a followup, since it's clear enough from the comments.

achow101 replied:
I would prefer to leave it as-is for now. That can be changed in a followup.

Sjors replied:
That's fine, but I'll keep stalking you :-)

Sjors commented on src/wallet/wallet.cpp:
```diff
@@ -703,7 +713,7 @@ bool CWallet::EncryptWallet(const SecureString& strWalletPassphrase)
         Unlock(strWalletPassphrase);
 
-        // if we are using HD, replace the HD seed with a new one
+        // if we are using HD, create a fresh seed or replace the existing seed with a new one
```
I don't see how "create a fresh seed" happens here, because if there's no seed `IsHDEnabled()` returns false.

achow101 replied:
Outdated comment. I've changed it back.

Sjors replied:
Might be useful to add "If no HD seed is set, e.g. because it is a blank wallet, do not add one.", can be done in the same followup as renaming `IsHDEnabled()`.

Sjors reviewed (approved):
utACK 707eaba, but please double check my "create a fresh seed" comment before merging.

I like the new `CanGenerateKeys` and `CanGetAddresses` approach, and the fact that the `blank` flag now really indicates the wallet is empty, rather than the more complicated condition we had before.

I checked again that v0.17.1 refuses to open a blank wallet, and that after `sethdseed` it does.

whitebreadapple commented:
Hacked accounts lisa isn't the only one responding 9705905818call her pleaseSent from my Verizon, Samsung Galaxy smartphone
null

whitebreadapple commented:
Hacked account

On Sun, Feb 10, 2019, 10:49 AM lisakepl84 <lisakepl84@gmail.com wrote:

> Hacked accounts lisa isn't the only one responding 9705905818call her
> please
>
>
>
> Sent from my Verizon, Samsung Galaxy smartphone
>

Sjors commented:
re-utACK 7687f78
